compiler/rustc_hir_analysis/src/lib.rs

```diff
@@ -127,20 +127,35 @@ fn require_c_abi_if_c_variadic(
     }
 
     let extended_abi_support = tcx.features().extended_varargs_abi_support();
-    let conventions = match (extended_abi_support, abi.supports_varargs()) {
-        // User enabled additional ABI support for varargs and function ABI matches those ones.
-        (true, true) => return,
-
-        // Using this ABI would be ok, if the feature for additional ABI support was enabled.
-        // Return CONVENTIONS_STABLE, because we want the other error to look the same.
-        (false, true) => {
-            feature_err(&tcx.sess, sym::extended_varargs_abi_support, span, UNSTABLE_EXPLAIN)
-                .emit();
-            CONVENTIONS_STABLE
+    let extern_system_varargs = tcx.features().extern_system_varargs();
+    let conventions = if extended_abi_support { CONVENTIONS_UNSTABLE } else { CONVENTIONS_STABLE };
+    match abi {
+        ExternAbi::System { .. } => {
+            // User enabled additional ABI support for varargs and function ABI matches those ones.
+            if extern_system_varargs {
+                return;
+            } else {
+                // Using this ABI would be ok, if the feature for additional ABI support was enabled.
+                feature_err(&tcx.sess, sym::extern_system_varargs, span, UNSTABLE_EXPLAIN).emit();
+            }
+        }
+        abi => {
+            if abi.supports_varargs() {
+                // User enabled additional ABI support for varargs and function ABI matches those ones.
+                if extended_abi_support {
+                    return;
+                } else {
+                    // Using this ABI would be ok, if the feature for additional ABI support was enabled.
+                    feature_err(
+                        &tcx.sess,
+                        sym::extended_varargs_abi_support,
+                        span,
+                        UNSTABLE_EXPLAIN,
+                    )
+                    .emit();
+                };
+            }
         }
-
-        (false, false) => CONVENTIONS_STABLE,
-        (true, false) => CONVENTIONS_UNSTABLE,
     };
 
     tcx.dcx().emit_err(errors::VariadicFunctionCompatibleConvention { span, conventions });
```

Comment:
Ideally we'd
- check if this is an actual error path (else bail)
- then compute the error we want to emit
- then emit one error

But also, ideally we'd not have to hardcode anything about the error we emit here: with https://github.com/rust-lang/rust/pull/136901 we'd be able to simply filter `ExternAbi::ALL_VARIANTS` to "everything that actually can support varargs", then format and emit the list.